### DISCUSSION TITLE
Extend Frontend / Backend. Which programming language was used for EMQX

### DISCUSSION THREAD

**Discussion by jpiwek** [emqx/emqx#13706] [Q&A]
Hi,

where to find  more details about how EMQX project (Front-, Backend) is structured and which programming language is used for both in-case I want to writte some add-ons?

**Comment by id** [comment]
Hi @jpiwek. EMQX is written in Erlang as you can see by this repository.
And the UI is React/Typescript: [emqx/emqx-dashboard5](https://github.com/emqx/emqx-dashboard5).
